src/app/reporting/LoggingManagement.cpp

```diff
@@ -0,0 +1,1306 @@
+    sInstance.mState        = kLoggingManagementState_Shutdown;
+    sInstance.mpEventBuffer = NULL;
+    Platform::CriticalSectionExit();
+}
+
+/**
+ * @brief Set the ExchangeManager to be used with this logging subsystem.  On some
+ *   platforms, this may need to happen separately from CreateLoggingManagement() above.
+ *
+ * @param[in] inMgr         ExchangeManager to be used with this logging subsystem
+ */
+CHIP_ERROR LoggingManagement::SetExchangeManager(Messaging::ExchangeManager * apMgr)
+{
+    mpExchangeMgr = apMgr;
+    return CHIP_NO_ERROR;
+}
+
+/**
+ * @brief
+ *   LoggingManagement constructor
+ *
+ * Initialize the LoggingManagement with an array of LogStorageResources.  The
+ * array must provide a resource for each valid priority level, the elements
+ * of the array must be in increasing numerical value of priority (and in
+ * increasing priority); the first element in the array corresponds to the
+ * resources allocated for least important events, and the last element
+ * corresponds to the most critical events.
+ *
+ * @param[in] apMgr         ExchangeManager to be used with this logging subsystem
+ *
+ * @param[in] aNumBuffers  Number of elements in inLogStorageResources array
+ *
+ * @param[in] apLogStorageResources  An array of LogStorageResources for each priority level.
+ *
+ */
+
+LoggingManagement::LoggingManagement(Messaging::ExchangeManager * apMgr, size_t aNumBuffers,
+                                     const LogStorageResources * const apLogStorageResources)
+{
+    CircularEventBuffer * current = NULL;
+    CircularEventBuffer * prev    = NULL;
+    CircularEventBuffer * next    = NULL;
+    size_t i, j;
+
+    VerifyOrDie(aNumBuffers > 0);
+
+    mpExchangeMgr = apMgr;
+
+    for (j = 0; j < aNumBuffers; j++)
+    {
+        i = aNumBuffers - 1 - j;
+
+        next = (i > 0) ? static_cast<CircularEventBuffer *>(apLogStorageResources[i - 1].mpBuffer) : NULL;
+
+        VerifyOrDie(apLogStorageResources[i].mBufferSize > sizeof(CircularEventBuffer));
+
+        new (apLogStorageResources[i].mpBuffer)
+            CircularEventBuffer(static_cast<uint8_t *>(apLogStorageResources[i].mpBuffer) + sizeof(CircularEventBuffer),
+                                (uint32_t)(apLogStorageResources[i].mBufferSize - sizeof(CircularEventBuffer)), prev, next);
+
+        current = prev                  = static_cast<CircularEventBuffer *>(apLogStorageResources[i].mpBuffer);
+        current->mProcessEvictedElement = AlwaysFail;
+        current->mAppData               = NULL;
+        current->mPriority              = apLogStorageResources[i].mPriority;
+        if ((apLogStorageResources[i].mCounterStorage != NULL) && (apLogStorageResources[i].mCounterKey != NULL) &&
+            (apLogStorageResources[i].mCounterEpoch != 0))
+        {
+
+            // We have been provided storage for a counter for this priority level.
+            new (apLogStorageResources[i].mCounterStorage) PersistedCounter();
```

Comment:
is placement new guaranteed properly aligned here?
Why are we using it and not passing in objects?

Reply:
From test application, it indeed provide storage, 
https://github.com/project-chip/connectedhomeip/blob/ea07b64987195a3b13663756f148951b17b3e2dc/src/app/tests/integration/chip_im_responder.cpp#L116
for example, gDebugEventBuffer has been saved into void * mpBuffer in LogStorageResources, we need to convert to CircularEventBuffer using the below new CircularEventBuffer and storage saved in mpBuffer(from gDebugEventBuffer)
        https://github.com/project-chip/connectedhomeip/blob/ea07b64987195a3b13663756f148951b17b3e2dc/src/app/reporting/LoggingManagement.cpp#L320

chip::PersistedCounter * mCounterStorage is also member in LogStorageResources, we use new to create this counter, which is kind of tracker.

As you see, the overall circular buffer storage is set by application, we hide circular buffer and counter buffer beneath user's view so that user just need provide the storage.

Reply:
Removed new placement. Done, thanks